structured/structured.go

```diff
@@ -303,3 +304,16 @@ func (desc *DatabaseDescriptor) Show() (UserPrivilegeList, error) {
 
 	return permissions.Show(), nil
 }
+
+// HasPrivilege returns true if `user` has `privilege` on this descriptor.
+func (desc *DatabaseDescriptor) HasPrivilege(user string, privilege parser.PrivilegeType) bool {
+	switch privilege {
+	case parser.PrivilegeRead:
+		result := sort.SearchStrings(desc.Read, user)
+		return result < len(desc.Read) && desc.Read[result] == user
+	case parser.PrivilegeWrite:
+		result := sort.SearchStrings(desc.Write, user)
+		return result < len(desc.Write) && desc.Write[result] == user
+	}
```

Comment:
I'd adjust this slightly so that the switch maps to a set of strings which is then searched outside of the switch:

```
  var users []string
  switch privilege {
  case parser.PrivilegeRead:
    users = desc.Read
  case parser.PrivilegeWrite:
    users = desc.Write
  default:
    return false
  }
  result := sort.SearchStrings(users, user)
  return result < len(users) && users[result] == user
```

This is slightly longer now, but will making adding additional privileges easier.